basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java

```diff
@@ -0,0 +1,156 @@
+import hudson.EnvVars;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.tasks.BuildWrapperDescriptor;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import jenkins.model.Jenkins;
+import jenkins.tasks.SimpleBuildWrapper;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+/**
+ * A step that runs a {@link SimpleBuildWrapper} as defined in Jenkins core.
+ */
+public class CoreWrapperStep extends AbstractStepImpl {
+
+    private final SimpleBuildWrapper delegate;
+
+    @DataBoundConstructor public CoreWrapperStep(SimpleBuildWrapper delegate) {
+        this.delegate = delegate;
+    }
+
+    public SimpleBuildWrapper getDelegate() {
+        return delegate;
+    }
+
+    public static final class Execution extends AbstractStepExecutionImpl {
+
+        private static final long serialVersionUID = 1;
+
+        @Inject(optional=true) private transient CoreWrapperStep step;
+        @StepContextParameter private transient Run<?,?> run;
+        @StepContextParameter private transient FilePath workspace;
+        @StepContextParameter private transient Launcher launcher;
+        @StepContextParameter private transient TaskListener listener;
+        // TODO ideally we would like to just create a fresh EnvVars with only our own bindings.
+        // But DefaultStepContext has no notion of merging multiple EnvVars instances across nested scopes.
+        // So we need to pick up the bindings created by ExecutorStepExecution and append to them.
+        // This has the unfortunate effect of “freezing” any custom values set via EnvActionImpl.setProperty for the duration of this step,
+        // because these will also be present in the inherited EnvVars.
+        @StepContextParameter private transient EnvVars env;
+
+        @Override public boolean start() throws Exception {
+            SimpleBuildWrapper.Context c = new SimpleBuildWrapper.Context();
+            step.delegate.setUp(c, run, workspace, launcher, listener, env);
+            BodyInvoker bodyInvoker = getContext().newBodyInvoker();
+            Map<String,String> overrides = c.getEnv();
+            if (!overrides.isEmpty()) {
+                EnvVars overridden = new EnvVars(env);
+                overridden.overrideExpandingAll(overrides);
+                bodyInvoker.withContext(overridden);
+            }
+            SimpleBuildWrapper.Disposer disposer = c.getDisposer();
+            bodyInvoker.withCallback(disposer != null ? new Callback(disposer) : BodyExecutionCallback.wrap(getContext())).start();
+            return false;
+        }
+
+        @Override public void stop(Throwable cause) throws Exception {
+            // should be no need to do anything special (but verify in JENKINS-26148)
+        }
+
+    }
+
+    private static final class Callback extends BodyExecutionCallback {
```

Comment:
`ExecutorStepExecution.Callback` does not fit the pattern. `TimeoutStepExecution.Callback` might, but maybe not, since it is not `static` (uses `TimeoutStepExecution.killer`). `WorkspaceStepExecution.Execution` and `TmpDirStepExecution.Execution` could share a common convenience superclass with `CoreWrapperStep.Callback`. There is not much boilerplate here but could make sense as a separate refactoring PR.